plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AbstractConfluentRowEncoder.java

```diff
@@ -0,0 +1,246 @@
+                checkState(rowType.getFields().size() == fieldSchema.getFields().size(), "Mismatch in number of fields in row and schema");
+                Block rowBlock = block.getObject(position, Block.class);
+                for (int index = 0; index < rowType.getFields().size(); index++) {
+                    RowType.Field field = rowType.getFields().get(index);
+                    Schema.Field avroField = fieldSchema.getFields().get(index);
+                    Schema avroFieldSchema = avroField.schema();
+                    Object value = getValue(field.getType(), rowBlock, index, avroFieldSchema);
+                    recordBuilder.set(avroField, value);
+                }
+            }
+            return recordBuilder.build();
+        }
+        else {
+            throw new UnsupportedOperationException(format("Unsupported type '%s'", type));
+        }
+    }
+
+    @VisibleForTesting
+    public static Schema extractBaseType(Schema schema)
+    {
+        requireNonNull(schema, "schema is null");
+        if (!schema.isUnion()) {
+            return schema;
+        }
+        Optional<Schema> baseSchema = schema.getTypes().stream()
+                .filter(fieldSchema -> fieldSchema.getType() != Schema.Type.NULL)
+                .map(nonNullSchema -> {
+                    if (nonNullSchema.isUnion()) {
+                        return extractBaseType(nonNullSchema);
+                    }
+                    return nonNullSchema;
+                }).findFirst();
+        checkState(baseSchema.isPresent(), "baseSchema is empty");
+        return baseSchema.get();
+    }
+
+    private boolean isEmptyStruct(Schema schema, Type type)
+    {
+        // If the avro schema is a record with no fields and the empty field strategy is ADD_DUMMY
+        checkState(schema.getType() == Schema.Type.RECORD, "Unexpected type '%s' for record schema", schema.getType());
+        checkState(type instanceof RowType, "Unexpected type '%s' for trino struct field", type.getTypeId());
+        if (schema.getFields().isEmpty() && getEmptyFieldStrategy(session) == ADD_DUMMY) {
+            RowType rowType = (RowType) type;
+            if (rowType.getFields().size() == 1) {
+                RowType.Field field = getOnlyElement(rowType.getFields());
+                if (field.getName().isPresent() && field.getName().get().equals(DUMMY_FIELD_NAME)) {
+                    return field.getType() instanceof BooleanType;
+                }
+            }
+        }
+        return false;
+    }
+
+    private void resetColumnIndex()
+    {
+        currentColumnIndex = 0;
+    }
+
+    protected int getCurrentColumnIndex()
+    {
+        return currentColumnIndex;
+    }
+
+    @Override
+    public byte[] toByteArray()
+    {
+        byte[] bytes = kafkaAvroSerializer.serialize(topic, buildRow());
+        resetColumnIndex();
+        return bytes;
+    }
+
+    @Override
+    public void close()
+    {
+        // Clear thread local encoder
```

Comment:
did you mean to add some code here?

Reply:
Removed - that was left over from an older version